src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -806,7 +807,9 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
         // worker exit to avoid process and resource leak.
         NotifyCoreWorkerToKillActor(actor, death_cause, force_kill);
       }
-      CancelActorInScheduling(actor, TaskID::ForActorCreationTask(actor_id));
+      if (!actor_scheduling_canceled) {
+        CancelActorInScheduling(actor, TaskID::ForActorCreationTask(actor_id));
```

Comment:
what's the side effect of calling `CancelActorInScheduling` if actor_scheduling_canceled is true?

Reply:
In gcs actor scheduler, `actor_scheduling_canceled` is set true only when gcs' `ClusterTaskManager` ReplyCancelled with `SCHEDULING_CANCELLED_UNSCHEDULABLE`.  If we call `CancelActorInScheduling` in this case, we will remove the task from `ClusterTaskManager`'s pending queue (via calling`GcsActorManager::RemovePendingActor()`). Then when we call `work_it = work_queue.erase(work_it);` (at line 122 in `ClusterTaskManager::ScheduleAndDispatchTasks()`), we will run into a `double free` fatal failure. 

To solve this problem, we can just call `work_it = work_queue.erase(work_it);` before calling `ReplyCancelled(*, rpc::RequestWorkerLeaseReply::SCHEDULING_CANCELLED_UNSCHEDULABLE, *)`. But then we call `GcsActorManager::RemovePendingActor()`, we will not find the task from the pending queue. Then we will go to `GcsActorScheduler::CancelOnLeasing()` and get a RAY_CHECK failure (line 229 in `GcsActorScheduler::CancelOnLeasing()`), because this worker lease has never been sent out.

Anyway, this logic is complicated, and I feel the current change is the most straightforward and easy-to-understand way.

Reply:
The part has been moved to `ClusterTaskManager::ScheduleAndDispatchTasks`, which makes it much simpler. @scv119